utils/yoke/src/cartable_ptr.rs

```diff
@@ -0,0 +1,395 @@
+    unsafe fn clone_raw(pointer: NonNull<T>) {
+        // Safety: The caller safety of this function says that:
+        // 1. The pointer was obtained through Arc::into_raw
+        // 2. The associated Arc instance is valid
+        // Further, this impl is not defined for anything but the global allocator.
+        Arc::increment_strong_count(pointer.as_ptr());
+    }
+}
+
+/// A type with similar semantics as `Option<C<T>>` but with a niche.
+///
+/// Will panic on types with alignment 1.
+#[derive(Debug)]
+pub struct CartableOptionPointer<C>
+where
+    C: CartablePointerLike,
+{
+    /// The inner pointer.
+    ///
+    /// # Invariants
+    ///
+    /// 1. Must be either `SENTINEL_PTR` or created from `CartablePointerLike::into_raw`
+    /// 2. If non-sentinel, must _always_ be for a valid SelectedRc
+    inner: NonNull<C::Raw>,
+    _cartable: PhantomData<C>,
+}
+
+impl<C> CartableOptionPointer<C>
+where
+    C: CartablePointerLike,
+{
+    /// Creates a new instance corresponding to a `None` value.
+    #[inline]
+    pub(crate) fn none() -> Self {
+        Self {
+            inner: sentinel_for::<C::Raw>(),
+            _cartable: PhantomData,
+        }
+    }
+
+    /// Creates a new instance corresponding to a `Some` value.
+    #[inline]
+    pub(crate) fn from_cartable(cartable: C) -> Self {
+        Self {
+            inner: cartable.into_raw(),
+            _cartable: PhantomData,
+        }
+    }
+
+    /// Returns whether this instance is `None`. From the return value:
+    ///
+    /// - If `true`, the instance is `None`
+    /// - If `false`, the instance is a valid `SelectedRc`
+    #[inline]
+    pub fn is_none(&self) -> bool {
+        self.inner == sentinel_for::<C::Raw>()
+    }
+}
+
+impl<C> Drop for CartableOptionPointer<C>
+where
+    C: CartablePointerLike,
+{
+    fn drop(&mut self) {
+        let ptr = self.inner;
+        if ptr != sentinel_for::<C::Raw>() {
+            // By the invariants, `ptr` is a valid raw value since it's
+            // either that or sentinel, and we just checked for sentinel.
+            // We will replace it with the sentinel and then drop `ptr`.
+            self.inner = sentinel_for::<C::Raw>();
+            // Safety: by the invariants, `ptr` is a valid raw value.
+            unsafe { C::drop_raw(ptr) }
+        }
+    }
+}
+
+impl<C> Clone for CartableOptionPointer<C>
+where
+    C: CloneableCartablePointerLike,
+{
+    fn clone(&self) -> Self {
+        let ptr = self.inner;
+        if ptr != sentinel_for::<C::Raw>() {
+            // By the invariants, `ptr` is a valid raw value since it's
+            // either that or sentinel, and we just checked for sentinel.
+            // Safety: by the invariants, `ptr` is a valid raw value.
+            unsafe { C::clone_raw(ptr) }
+        }
+        Self {
+            inner: self.inner,
+            _cartable: PhantomData,
+        }
+    }
+}
+
+// Safety: logically an Option<C>. Has same bounds as Option<C>
```

Comment:
nit:  with my other PR landing you can potentially relate to the invariants specified there